RecoEgamma/EgammaTools/interface/EgammaPCAHelper.h

```diff
@@ -0,0 +1,115 @@
+//--------------------------------------------------------------------------------------------------
+//
+// EGammaPCAHelper
+//
+// Helper Class to compute PCA
+//
+//
+//--------------------------------------------------------------------------------------------------
+#ifndef EGammaPCAHelper_H
+#define EGammaPCAHelper_H
+
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+
+#include "DataFormats/CaloRecHit/interface/CaloCluster.h"
+#include "DataFormats/HGCRecHit/interface/HGCRecHitCollections.h"
+#include "DataFormats/ForwardDetId/interface/HGCalDetId.h"
+#include "DataFormats/Math/interface/Vector3D.h"
+#include "DataFormats/ParticleFlowReco/interface/HGCalMultiCluster.h"
+
+#include "RecoEgamma/EgammaTools/interface/Spot.h"
+#include "RecoEgamma/EgammaTools/interface/LongDeps.h"
+#include "RecoEgamma/EgammaTools/interface/ShowerDepth.h"
+#include "RecoLocalCalo/HGCalRecAlgos/interface/RecHitTools.h"
+#include "FastSimulation/CaloGeometryTools/interface/Transform3DPJ.h"
+#include <map>
+
+#include "TPrincipal.h"
+
+class HGCalRecHit;
+
+class EGammaPCAHelper
+{
+public:
+    typedef ROOT::Math::Transform3DPJ Transform3D;
+    typedef ROOT::Math::Transform3DPJ::Point Point;
+
+    EGammaPCAHelper();
+    ~EGammaPCAHelper();
+
+    // for the GsfElectrons
+    void storeRecHits(const reco::CaloCluster & theCluster );
+    void storeRecHits(const reco::HGCalMultiCluster &cluster );
+
+    const TPrincipal & pcaResult();
+    /// to set from outside - once per event
+    void setHitMap( std::map<DetId,const HGCRecHit *> * hitMap) ;
+    /// to compute from inside - once per event
+    void fillHitMap(const HGCRecHitCollection & HGCEERecHits,
+                    const HGCRecHitCollection & HGCFHRecHits,
+                    const HGCRecHitCollection & HGCBHRecHits);
+
+    std::map<DetId,const HGCRecHit *> * getHitMap(){return hitMap_;}
+
+    void setRecHitTools(const hgcal::RecHitTools * recHitTools );
+
+    inline void setdEdXWeights(const std::vector<double> & dEdX){ dEdXWeights_ = dEdX;}
+
+    void pcaInitialComputation() {
+        computePCA(-1.,false);
+    }
+
+    void computePCA(float radius, bool withHalo=true);
+    const math::XYZPoint  & barycenter() const {return barycenter_;}
+    const math::XYZVector & axis() const {return axis_;}
+
+    void computeShowerWidth(float radius, bool withHalo=true);
+
+    inline double sigmaUU() const { return checkIteration()? sigu_ : -1. ;}
+    inline double sigmaVV() const { return checkIteration()? sigv_ : -1. ;}
+    inline double sigmaEE() const { return checkIteration()? sige_ : -1. ;}
+    inline double sigmaPP() const { return checkIteration()? sigp_ : -1. ;}
+
+    inline TVectorD eigenValues () const {return checkIteration()? *pca_->GetEigenValues() : TVectorD(-1.,-1.,-1.);}
+    inline TVectorD sigmas() const {return checkIteration()? *pca_->GetSigmas() : TVectorD(-1.,-1.,-1.);}
```

Comment:
this can return by reference or by const pointer as the pca interface allows.

Reply:
Can't be ref, as reference to temporary in case `checkIteration()` is false.  Anyway, its inlined, so the function call doesn't even exist...
Made it const though.

Reply:
It's still a copy of something pretty large.
A ref can be made to a static member.
Do you have evidence from compiled code that there is no copy?

Reply:
Its not large, always length 3 because the PCA has 3 dimensions

Reply:
The "size 3" is actually about 100 bytes, which is large in this context.

//from TVectorT
   Int_t    fNrows;                // number of rows
   Int_t    fRowLwb;               // lower bound of the row index
   Element *fElements;             //[fNrows] elements themselves
   Element  fDataStack[kSizeMax];  //! data container
   Bool_t   fIsOwner;              //!default kTRUE, when Use array kFALSE
//from TObject
   UInt_t         fUniqueID;   //object unique identifier
   UInt_t         fBits;       //bit field status word
.. and a vtable pointer for inheritance

Reply:
Ok fine.  In fact I discovered that the `TVectorD` constructor is absolutely bonkers:
```
$ root -l
root [0] TVectorD asdf(-1., -1., -1.);
Error in <TVectorT<double>::TVectorT(Int_t, Int_t, ...)>: argument list must be terminated by "END"
root [1] asdf.Print()

Vector (1)  is as follows

     |        1  |
------------------
  -1 |-1

root [2] asdf(-1)
(double) -1.000000
root [3] asdf(0)
Error in <TVectorT<double>::operator()>: Request index(0) outside vector range of -1 - 0
(double) nan
```
so we'll let the function return the (always valid) reference to the `TVectorD` that lives inside `TPrincipal`
@artlbv you might find this humorous.